### ISSUE TITLE
Unable to get this extension to work

### ISSUE THREAD

**Issue by cnokello** [vincentdieltiens/WebSocketFileTransfer#1]
This sounds like a great extension. However, I've tried to use it with no success so far. I can see the file details on the socket server side, but I'm not sure how I can access the actual file. 

On the client side, the following exception is thrown:

'Uncaught TypeError: Cannot read property 'size' of undefined'

On debugging, this error is thrown in the 'socket.onmessage()' method in WebSocketFileTransfer.js:137.

I'm using this file upload widget, https://github.com/ivan-novakov/extjs-upload-widget,  by @ivan.

The relevant section of the code is the part that does the actual uploading:

```
 uploadItem : function(item) {
       var file = item.getFileApiObject();
       if (!file) {
           return;
       }

       item.setUploading();

       console.log(file);

       if(!WebSocketFileTransfer.supported()) {
        console.log('WebSocket File Transfer is not fully supported by your browser.');
        return;
       }
       var transfer = new WebSocketFileTransfer({
       url: 'ws://localhost:8088/ws',
       file: file,
       blockSize: 1024,
       type: WebSocketFileTransfer.binarySupported() ? 'binary' : 'base64',

       progress: function(event) {
        console.log('File upload in progress');
       },

       success: function(event) {
        console.log('File successfully uploaded.');
       }

    });

    transfer.start();

},
```

The original functionality for uploading a file before my changes can be found at https://github.com/ivan-novakov/extjs-upload-widget/blob/master/lib/upload/uploader/ExtJsUploader.js.

You also mentioned a wrapper class. If you care to expound on that, please.

'item' uses the native File API. Have a look at https://github.com/ivan-novakov/extjs-upload-widget/blob/master/lib/upload/Item.js.

I can access the file details (name, size, etc) on the socket server side. But I ain't sure how I'm to get the file itself. I'm using Netty socket server.

**Comment by vincentdieltiens** [comment]
You can see in my server example (in php) here: https://github.com/vincentdieltiens/PhpWebSocketFileTransferServer/blob/master/upload_server.php

These are the steps :
1. when calling the `start()` method, the WebSocketFileTransfer client establishes the WebSocket connection with the server.
2. The server receives the connection and return an ACK (according to the WebSocket protocol). This is handled by any Websocket library on the server
3. When the ACK is received from server, the WebSocketFileTransfer sends The AUTH command to the server if you define a login/password couple via the `start` method. If no login/password defined, go to the step 5 directly.
4. The server receive the AUTH command, tests if the login/password is good and answer according to the result.
   (No handled by my sample server)
   - If login is good :

``` json
{
    "type": "AUTH",
    "message": "Authentification success",
    "code": 200
}
```
- If login is not good :

``` json
{
    "type": "AUTH",
    "message": "Authentification failed",
    "code": 403
}
```
1. The WebSocketFileTransfer sends the STOR command to the server
2. The Server get the upload informations (file name, file size, ...), creates a blank file and returns a json object to the client :

``` json
{
    "type": "STOR",
    "message": "Upload initialized. Wait for data",
    "code": 200
}
```
1. When receiving the STOR response from the server with a code 200, the WebSocketFileTransfer starts sending the file in binary or base64 according to the option you defined with the DATA command
2. The server receives the first piece of data and append it to the blank file and return a json object :

``` json
{
    "type": "DATA",
    "code": 200,
    "bytesRead": numberOfBytesReceivedAndReadByTheServer
}
```
1. At each step a DATA response is received by WebSocketFileTransfer, it send another piece of data, and so on.
2. After all the file is sent, the WebSocketFileTransfer close the connection to the server
3. When the connection is closed, the server close the file and save it somewhere on the server.

Hope it is helping ;-)

**Comment by cnokello** [comment]
Thanks a lot, Vincent. 
This is very informative and well put down. There are some steps I didn't implement; the DATA ones. I'll incorporate them and then let you know what happens.

Kind Regards.

Okello Nelson

[edit vincentdieltiens : remove cnkello's signature for better reading]

**Comment by vincentdieltiens** [comment]
See the diagram I put yesterday ;-)

**Comment by vincentdieltiens** [comment]
Is it working now ?

**Comment by cnokello** [comment]
Hi Vincent,

I'm sorry I got swamped that I forgot to get back to you for your assistance. Yes, I got the extension to work. I'll post details of what I did in due course. Thanks a lot.

Kind Regards.

Okello Nelson

E: me@okellonelson.com

M: +254 722 137 826

Twitter: @cnokello

W: http://cnokello.blogspot.com | http://okellonelson.com | http://enterface.co.ke

Date: Mon, 13 May 2013 00:28:44 -0700
